esda/shape.py

```diff
@@ -0,0 +1,376 @@
+import pygeos
+import geopandas, pandas
+import numpy
+from numba import njit, prange
+
+# -------------------- UTILITIES --------------------#
+def _cast(collection):
+    """
+    Cast a collection to a pygeos geometry array.
+    """
+    if isinstance(collection, (geopandas.GeoSeries, geopandas.GeoDataFrame)):
+        return collection.values.data.squeeze()
+    elif pygeos.is_geometry(collection).all():
+        return collection
+    return pygeos.from_shapely(collection).squeeze()
+
+
+def get_angles(collection, return_indices=False):
```

Comment:
I either don't understand how this should return the results or it has a bug. Consider the following example:

```py
from shapely import geometry
from pygeos import from_shapely
from esda.shape import *
from numpy import testing, array

shape = from_shapely([
            geometry.Polygon(
                [
                    (0, 0),
                    (0, 0.5),
                    (1.25, 0),
                    (0.5, 0),
                    (0.3, 0),
                ]
            ) for _ in range(5)]
        )

angles, ind = get_angles(shape, return_indices=True)
pandas.Series(angles).groupby(ind).sum()

0    9.424778
1    4.712389
2    4.712389
3    4.712389
4    6.283185
dtype: float64

angles
array([1.19028995, 0.38050638, 3.14159265, 3.14159265, 1.57079633,
       0.        , 1.19028995, 0.38050638, 3.14159265, 0.        ,
       0.        , 0.        , 1.19028995, 0.38050638, 3.14159265,
       3.14159265, 0.        , 0.        , 0.78539816, 0.78539816,
       3.14159265, 3.14159265, 0.        , 0.        , 0.        ])

ind
array([0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 4, 4,
       4, 4, 4])
```

All five polygons are identical. However, reported angles are not equal. How is this supposed to work? I wanted to use it to identify which vertices capture a corner with an angle within a specific threshold but I can't wrap my head around this output.

Reply:
Concerned.... I'll dig into this. I think it may be a trigonometric thing, but I'm not sure why the results would change for identical shapes.

There may be more efficient ways to compute the number of reflexive angles in a shape, too, which I'm not aware of.

Reply:
Oddly enough, `[get_angles(s, return_indices=True) for s in shape]`gives consistent results. Thus, it may be something about how I'm attempting to handle multi-part geometries. Will investigate further.

Reply:
It's an off-by-one error in the construction of the offset in `_get_angles()`.